File: 238-product-of-array-except-self/product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> prefix(nums.size(), 1);
        for(int i=1; i<nums.size(); i++){
            prefix[i] = prefix[i-1] * nums[i-1];
        }
        int product = nums[nums.size()-1];
        for(int i=nums.size()-2; i>=0; i--){
            prefix[i] *= product;
            product *= nums[i];
        }
        return prefix;
    }
};

File: 238-product-of-array-except-self/product-of-array-except-self.go
func productExceptSelf(nums []int) []int {
    prod := make([]int, len(nums))
    product := 1
    for i := range nums {
        prod[i] = product
        product *= nums[i]
    }

    product = 1
    for i := len(nums)-1; i >= 0; i-- {
        prod[i] *= product
        product *= nums[i]
    }

    return prod
}